pkg/auth/providers/keycloakoidc/keycloak_client.go

```diff
@@ -0,0 +1,179 @@
+package keycloakoidc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+type searchResults struct {
+	Items []Account `json:"items"`
+}
+
+//Account defines properties an account in keycloak has
+type Account struct {
+	ID            int    `json:"id,omitempty"`
+	Email         string `json:email,omitempty`
+	EmailVerified string `json:emailVerified,omitempty`
+	Username      string `json:"username,omitempty"`
+	Enabled       bool   `json:"enabled,omitempty"`
+	Name          string `json:"firstName,omitempty"`
+	LastName      string `json:"lastName,omitempty"`
+	Type          string
+}
+
+//Group defines properties a group in keycloak has
+type Group struct {
+	ID        int     `json:"id,omitempty"`
+	Name      string  `json:"name,omitempty"`
+	Subgroups []Group `json:subGroups,omitempty`
+}
+
+//KClient implements a httpclient for keycloak
+type KClient struct {
+	httpClient *http.Client
+}
+
+func (k *KClient) searchPrincipals(searchTerm, principalType string, accessToken string, config *v32.OIDCConfig) ([]Account, error) {
+	var accounts []Account
+
+	if principalType == "" || principalType == UserType {
+		var userAccounts []Account
+		searchURL := fmt.Sprintf("%s/%ss?search=%s", k.getSearchURL(config.Issuer), UserType, searchTerm)
+		search := URLEncoded(searchURL)
+
+		b, err := k.getFromKeyCloak(accessToken, search)
+		if err != nil {
+			// no match on search returns an error. do not log
+			return nil, nil
+		}
+		if err := json.Unmarshal(b, &userAccounts); err != nil {
+			logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+			return accounts, err
+		}
+		for _, u := range userAccounts {
+			u.Type = UserType
+			accounts = append(accounts, u)
+		}
+	}
+	if principalType == "" || principalType == GroupType {
+		var groups []Group
+		searchURL := fmt.Sprintf("%s/%ss?search=%s", k.getSearchURL(config.Issuer), GroupType, searchTerm)
+		search := URLEncoded(searchURL)
+
+		b, err := k.getFromKeyCloak(accessToken, search)
+		if err != nil {
+			// no match on search returns an error. do not log
+			return nil, nil
+		}
+		if err := json.Unmarshal(b, &groups); err != nil {
+			logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+			return accounts, err
+		}
+		for _, g := range groups {
+			accounts = append(accounts, Account{ID: g.ID, Name: g.Name, Type: GroupType})
+			subGroups := getSubGroups(g)
+			for _, sg := range subGroups {
+				accounts = append(accounts, Account{ID: sg.ID, Name: sg.Name, Type: GroupType})
+			}
+		}
+	}
+	return accounts, nil
+}
+
+func getSubGroups(group Group) []Group {
+	var groups []Group
+	if len(group.Subgroups) > 0 {
+		for _, sub := range group.Subgroups {
+			groups = append(groups, sub)
+			groups = append(groups, getSubGroups(sub)...)
```

Comment:
Does the above ""getFromKeyCloak search request return all group detail needed for all nested groups? Is there any limit/level upto which subgroups are listed back?

Reply:
Shouldn't getSubGroups(sub) call the provider to search for the subgroup attributes?

Reply:
Seems like we need some limit configured for this recursive call, is there a max level of nested groups set by keycloak? If not maybe we should add a default limit in our config

Reply:
Yeah I was also wondering about this when I coded it - I checked in the google provider to get an idea for a default amount and I didn't see one in there for the subgroup call - I did not kind any keycloak documentation with a limit identified for how many subgroups are possible either. Is there a specific default amount in mind that we think would be good?